### PR TITLE
Added TaintedBrokersSelector to kafkaClusterSpec

### PR DESCRIPTION
**Pull request by azun**
| Q               | A      |
| --------------- | ------ |
| Bug fix?        | no |
| New feature?    | yes |
| API breaks?     | no |
| Deprecations?   | no |
| Related tickets |  |
| License         | Apache 2.0 |

### What's in this PR?
Added label selector for broker pods that need to be recycled during reconcile

### Why?
Labeling pods for recycle can be used for rolling restarts or to offload rescheduling of the pods to the operator when nodes are tainted

### Checklist

- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)

### CODE REVIEW

**Review by alungu — APPROVED** [review]

File: pkg/resources/kafka/kafka.go [hunk 1]
```diff
@@ -880,6 +883,20 @@ func (r *Reconciler) handleRollingUpgrade(log logr.Logger, desiredPod, currentPo
 	return nil
 }
 
+func (r *Reconciler) isPodTainted(log logr.Logger, pod *corev1.Pod) bool {
```
Comment (amuraru) [inline pkg/resources/kafka/kafka.go:886]: Add docs here please to explain what this method does

File: pkg/resources/kafka/kafka.go [hunk 2]
```diff
@@ -779,6 +780,8 @@ func (r *Reconciler) handleRollingUpgrade(log logr.Logger, desiredPod, currentPo
 	switch {
 	case err != nil:
 		log.Error(err, "could not match objects", "kind", desiredType)
+	case r.isPodTainted(log, currentPod):
+		log.Info("pod has tainted labels", "pod", currentPod)
```
Comment (amuraru) [inline pkg/resources/kafka/kafka.go:784]:
```suggestion
		log.Info("pod has tainted labels, recreating it", "pod", currentPod)
```

File: config/base/crds/kafka.banzaicloud.io_kafkatopics.yaml [hunk 3]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.8.0
+    controller-gen.kubebuilder.io/version: v0.9.2
```
Comment (amuraru) [inline config/base/crds/kafka.banzaicloud.io_kafkatopics.yaml:6]: not part of this patch right?

**Comment by amuraru** [comment]
Please rebase our master branch first against banzaicloud/koperator 
https://github.com/banzaicloud/koperator/commit/e5722ba328678d1eda4045d5820c8e32bc62e208 is already there
and keep only the changes related to this PR

File: config/base/crds/kafka.banzaicloud.io_kafkatopics.yaml (hunk 3 above)
Comment (ecojan) [reply config/base/crds/kafka.banzaicloud.io_kafkatopics.yaml:6]: I noticed upstream was updated to this version, had the same curiosity.
Version comes from configured kubebuilder CONTROLLER_GEN_VERSION : https://github.com/adobe/koperator/blob/6396af5b5751e88637599e6c7e8c97b39d41cf87/Makefile#L20

It seems that this was a change that probably escaped the initial dependencies update.

**Review by ecojan — APPROVED** [review]
Outside of comments from the other review, this looks good!

File: api/v1beta1/zz_generated.deepcopy.go [hunk 4]
```diff
@@ -22,7 +22,8 @@ package v1beta1
 import (
 	networkingv1beta1 "github.com/banzaicloud/istio-client-go/pkg/networking/v1beta1"
 	metav1 "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
-	"k8s.io/api/core/v1"
+	corev1 "k8s.io/api/core/v1"
```
Comment (alungu) [inline api/v1beta1/zz_generated.deepcopy.go:25]: Could you double check if this change is needed, please?

File: pkg/resources/kafka/kafka.go (hunk 1 above)
Comment (amuraru) [reply pkg/resources/kafka/kafka.go:951]: Thanks

File: pkg/resources/kafka/kafka.go [hunk 5]
```diff
@@ -838,11 +839,13 @@ func (r *Reconciler) handleRollingUpgrade(log logr.Logger, desiredPod, currentPo
...
 		desiredPod.Spec.Tolerations = uniqueTolerations
 	}
-	// Check if the resource actually updated
+	// Check if the resource actually updated or if labels match TaintedBrokersSelector
 	patchResult, err := patch.DefaultPatchMaker.Calculate(currentPod, desiredPod)
 	switch {
 	case err != nil:
 		log.Error(err, "could not match objects", "kind", desiredType)
+	case r.isPodTainted(log, currentPod):
+		log.Info("pod has tainted labels, deleting it", "pod", currentPod)
```
Comment (ctrlaltluc) [inline pkg/resources/kafka/kafka.go:848]: Could there be situations where we need to handle here the case of another pod is already being terminated or still being started?

**Review by ctrlaltluc — APPROVED** [review]

File: api/v1beta1/zz_generated.deepcopy.go (hunk 4 above)
Comment (amuraru) [reply api/v1beta1/zz_generated.deepcopy.go:25]: @azun can you check this please?

**Comment by amuraru** [comment]
@azun can you please move forward this PR?

File: api/v1beta1/zz_generated.deepcopy.go (hunk 4 above)
Comment (azun) [reply api/v1beta1/zz_generated.deepcopy.go:25]: This is generated code. I'm not sure why it's adding `k8s.io/apimachinery/pkg/apis/meta/v1` which forces the alias for corev1

File: pkg/resources/kafka/kafka.go (hunk 5 above)
Comment (azun) [reply pkg/resources/kafka/kafka.go:848]: I piggybacked on the existing flow which was already checking pod states and change deltas before reconciling a pod.
